File: examples/benchmark_graph.py
import torch
import pandas as pd
from funcs_sigma import *
from parquetAD import FeynmanDiagram, load_leaf_info
import os
import torch.utils.benchmark as benchmark
import vegas
from vegas_torch import VegasMap
import numpy as np

enable_cuda = True
device = torch.device("cuda" if torch.cuda.is_available() and enable_cuda else "cpu")

root_dir = os.path.join(os.path.dirname(__file__), "source_codeParquetAD/")
num_loops = [2, 6, 15, 39, 111, 448]
order = 1
dim = 4 * order - 1
beta = 10.0
batch_size = 10000
Neval = 10

partition = [(order, 0, 0)]
name = "sigma"
df = pd.read_csv(os.path.join(root_dir, f"loopBasis_{name}_maxOrder6.csv"))
with torch.no_grad():
    # loopBasis = torch.tensor([df[col].iloc[:maxMomNum].tolist() for col in df.columns[:num_loops[order-1]]]).T
    loopBasis = torch.Tensor(df.iloc[: order + 1, : num_loops[order - 1]].to_numpy())
leafstates = []
leafvalues = []

for key in partition:
    key_str = "".join(map(str, key))
    state, values = load_leaf_info(root_dir, name, key_str)
    leafstates.append(state)
    leafvalues.append(values)

# for batchsize in [10**i for i in range(0, 7)]:
diagram = FeynmanDiagram(order, loopBasis, leafstates[0], leafvalues[0], batch_size)


# Vegas
@vegas.batchintegrand
def func(x):
    return torch.Tensor.numpy(diagram.prob(torch.Tensor(x)))


integration_domain = [[0, 1]] * dim
map_torch = VegasMap(diagram, dim, integration_domain, batch_size)
map_torch = map_torch.to(device)

var = torch.rand(batch_size, dim, device=device)
t0 = benchmark.Timer(
    stmt="map_torch.forward(var)",
    globals={"map_torch": map_torch, "var": var},
    label="Self-energy diagram (order {0} beta {1})".format(order, beta),
    sub_label="sampling using VEAGS map",
)

nfm = torch.load("nfm_o{0}_beta{1}.pt".format(order, beta))
nfm.eval()
nfm.p = diagram
nfm = nfm.to(device)


t1 = benchmark.Timer(
    stmt="nfm.forward(var)",
    globals={"nfm": nfm, "var": var},
    label="Self-energy diagram (order {0} beta {1})".format(order, beta),
    sub_label="sampling using normalizing flow",
)

t2 = benchmark.Timer(
    stmt="nfm.p.prob(var)",
    globals={"nfm": nfm, "var": var},
    label="Self-energy diagram (order {0} beta {1})".format(order, beta),
    sub_label="Evaluating integrand",
)

print(t0.timeit(Neval))
print(t1.timeit(Neval))
print(t2.timeit(Neval))

File: examples/benchmark_graph.jl
using ElectronLiquid, FeynmanDiagram
using BenchmarkTools
using MCIntegration, CSV, DataFrames

dim = 3
rs = 2.0
order = 3
mass2 = 0.5
beta = 10.0
isLayered2D = false
root_dir = "./source_codeParquetAD/"

para = UEG.ParaMC(rs=rs, beta=beta, Fs=-0.0, order=order, mass2=mass2, dim=dim)

partition = [(order, 0, 0)]
name = "sigma"
maxMomNum = order + 1
β, kF = para.dim, para.β, para.kF
kgrid = [kF]
ngrid = [0]
alpha = 3.0

df = CSV.read(root_dir * "loopBasis_$(name)_maxOrder6.csv", DataFrame)
loopBasis = [df[!, col][1:maxMomNum] for col in names(df)]
momLoopPool = FrontEnds.LoopPool(:K, dim, loopBasis)

# Helper function to convert string to integer vector
function _StringtoIntVector(str::AbstractString)
    pattern = r"[-+]?\d+"
    return [parse(Int, m.match) for m in eachmatch(pattern, str)]
end


extT_labels = [[[1, 1], [1, 2], [1, 3]]]
leafstates = Vector{Vector{Propagator.LeafStateAD}}()
leafvalues = Vector{Vector{Float64}}()
for key in partition
    key_str = join(string.(key))
    df = CSV.read(root_dir * "leafinfo_$(name)_$key_str.csv", DataFrame)
    leafstates_par = Vector{Propagator.LeafStateAD}()
    for row in eachrow(df)
        push!(leafstates_par, Propagator.LeafStateAD(row[2], _StringtoIntVector(row[3]), row[4:end]...))
    end
    push!(leafstates, leafstates_par)
    push!(leafvalues, df[!, names(df)[1]])
end

root = zeros(Float64, maximum(length.(extT_labels)))
# root = zeros(Float64, 3)
K = MCIntegration.FermiK(dim, kF, 0.5 * kF, 10.0 * kF, offset=1)
K.data[:, 1] .= 0.0
K.data[1, 1] = kF
# T = MCIntegration.Continuous(0.0, β; grid=collect(LinRange(0.0, β, 1000)), offset=1, alpha=alpha)
T = Continuous(0.0, β; alpha=alpha, adapt=true, offset=1)
T.data[1] = 0.0
X = MCIntegration.Discrete(1, 1, alpha=alpha)
ExtKidx = MCIntegration.Discrete(1, 1, alpha=alpha)

dof = [[order, order - 1, 1, 1]] # K, T, X, ExtKidx
# observable of sigma diagram of different permutations
obs = [zeros(ComplexF64, 1, 1) for _ in 1:length(dof)]
vars = (K, T, X, ExtKidx)

config = Configuration(;
    var=vars,
    dof=dof,
    type=ComplexF64, # type of the integrand
    obs=obs,
    userdata=(para, kgrid, ngrid, maxMomNum, extT_labels, leafstates, leafvalues, momLoopPool, root, isLayered2D, partition)
)


function rand_vars(vars)
    vars[1].data[:, 2:end] .= randn(size(vars[1].data[:, 2:end]))
    vars[2].data[2:end] .= rand(size(vars[2].data[2:end])[1])
end

Sigma.integrandKW_Clib(1, vars, config)

@btime Sigma.integrandKW_Clib(1, vars, config)

@benchmark Sigma.integrandKW_Clib(1, vars, config) setup = (rand_vars(vars))
println("Finished benchmarking Sigma.integrandKW_Clib")